source/base/Allocator.ooc

```diff
@@ -0,0 +1,36 @@
+/* This file is part of magic-sdk, an sdk for the open source programming language magic.
+ *
+ * Copyright (C) 2016 magic-lang
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+import Owner
+
+AbstractAllocator: abstract class {
+	_owner: Owner
+	free: func ~withCriteria (owner: Owner) {
+		if (owner == this _owner)
+			this free()
+	}
+	allocate: abstract func (size: SizeT) -> Pointer
+	deallocate: abstract func (pointer: Pointer)
+}
+
+MallocAllocator: class extends AbstractAllocator {
+	init: func (owner := Owner Receiver) { this _owner = owner }
+	allocate: override func (size: SizeT) -> Pointer { malloc(size) }
+	deallocate: override func (pointer: Pointer) { memfree(pointer) }
+}
+
+Allocator: abstract class {
+	_defaultAllocator: static AbstractAllocator = null
```

Comment:
Yes, we could have 
```ooc
This _defaultAllocator = This mallocAllocator
```
but I don't know where to put this as it depends on the order of initialization of static variables.
The default allocator is meant to be overriden as 
```ooc
Allocator defaultAllocator = myAllocator
```